locator/network_topology_strategy.cc

```diff
@@ -56,10 +56,11 @@ network_topology_strategy::network_topology_strategy(replication_strategy_params
             continue;
         }
 
-        if (boost::iequals(key, "replication_factor")) {
-            throw exceptions::configuration_exception(
-                "replication_factor is an option for SimpleStrategy, not "
-                "NetworkTopologyStrategy");
+        if (boost::equals(key, "replication_factor")) {
+            on_internal_error(rslogger, "replication_factor should have been replaced with a DC:RF mapping by now");
+        } else if (boost::iequals(key, "replication_factor")) {
+            throw exceptions::configuration_exception(format(
+                "'{}' is not a valid option, did you mean (lowercase) 'replication_factor'?", key));
```

Comment:
I've found no description of whether this key should be case-sensitive or not in docs :( I would say that this check should be made earlier, in the ks_prop_defs.cc::prepare_options() and here the strategy should just check `boost::iequals` (with `i`), as it used to

Reply:
> I've found no description of whether this key should be case-sensitive or not in docs

@avikivity [mentioned that](https://github.com/scylladb/scylladb/pull/16117#issuecomment-1821152441) + per https://github.com/scylladb/scylladb/issues/15336#issuecomment-1818921538 Cassandra returns an error if it's not lowercase as well.

> I would say that this check should be made earlier, in the ks_prop_defs.cc::prepare_options()

It seems to me that `ks_prop_defs.cc::prepare_options()` is called later than the check we're discussing here, because we first check for `replication_factor` presence (this check here) and later unroll it (the check you mentioned), isn't it correct?

> and here the strategy should just check boost::iequals (with i), as it used to

I assume that'd be true, given we're allowing for this tag to be case insensitive? But I doubt it's correct, as I wrote at the beginning.

Reply:
> I've found no description of whether this key should be case-sensitive or not in docs

@avikivity [mentioned that](https://github.com/scylladb/scylladb/pull/16117#issuecomment-1821152441) + per https://github.com/scylladb/scylladb/issues/15336#issuecomment-1818921538 Cassandra returns an error if it's not lowercase as well.

> I would say that this check should be made earlier, in the ks_prop_defs.cc::prepare_options()

It seems to me that `ks_prop_defs.cc::prepare_options()` is called later than the check we're discussing here, because we first check for `replication_factor` presence (this check here) and later unroll it (the check you mentioned), isn't it correct?

> and here the strategy should just check boost::iequals (with i), as it used to

I assume that'd be true, given we're allowing for this tag to be case insensitive? But I doubt it's correct, as I wrote at the beginning.

Reply:
How come the `ks_prop_defs.cc::prepare_options()` is called _later_? If it's called later, than benign "replication_factor" option remains in the options and this place should warn us that it's wrong to use it :exploding_head:

Reply:
I attached gdb to scylla, set breakpoints to both these places
```
(gdb) info breakpoints
Num     Type           Disp Enb Address            What
1       breakpoint     keep y   0x000000000587ca94 in cql3::statements::prepare_options(seastar::basic_sstring<char, unsigned int, 15u, true> const&, locator::token_metadata const&, std::map
<seastar::basic_sstring<char, unsigned int, 15u, true>, seastar::basic_sstring<char, unsigned int, 15u, true>, std::less<seastar::basic_sstring<char, unsigned int, 15u, true> >, std::allocat
or<std::pair<seastar::basic_sstring<char, unsigned int, 15u, true> const, seastar::basic_sstring<char, unsigned int, 15u, true> > > >, std::map<seastar::basic_sstring<char, unsigned int, 15u
, true>, seastar::basic_sstring<char, unsigned int, 15u, true>, std::less<seastar::basic_sstring<char, unsigned int, 15u, true> >, std::allocator<std::pair<seastar::basic_sstring<char, unsig
ned int, 15u, true> const, seastar::basic_sstring<char, unsigned int, 15u, true> > > > const&) at ./cql3/statements/ks_prop_defs.cc:31
        breakpoint already hit 1 time
2       breakpoint     keep y   0x00000000071ec231 in locator::network_topology_strategy::network_topology_strategy(locator::replication_strategy_params)
                                                   at ./locator/network_topology_strategy.cc:41
        breakpoint already hit 2 times
```
and executed:
```
cqlsh> CREATE KEYSPACE IF NOT EXISTS ks2 WITH REPLICATION = {'class': 'NetworkTopologyStrategy', 'Replication_factor': 3};
```
and the order of executing was this:
1. network_topology_strategy::network_topology_strategy
2. ks_prop_defs.cc::prepare_options
3. network_topology_strategy::network_topology_strategy

and my explanation to the above is:
1. here validate if tags passed to NTS are legit
2. here we unroll the `replication_factor` tag of NTS
3. probably someone created NTS 2nd time during CREATE KS execution, but `replication_factor` tag has been unrolled already, hence we're not entering `ks_prop_defs.cc::prepare_options` for the 2nd time

Reply:
Still this doesn't answer my question. The network_topology_strategy constructor has explicit check

```c++
    network_topology_strategy::network_topology_strategy(replication_strategy_params params) :
        ...
        if (boost::iequals(key, "replication_factor")) {
            throw exceptions::configuration_exception(
                "replication_factor is an option for SimpleStrategy, not "
                "NetworkTopologyStrategy");
        }
```

So ks-prop-defs must have replaced it already. Here's how it looks like when executing CREATE KEYSPACE statement

```
INFO  ... [shard 0:stmt] replication_strategy - network_topology_strategy() with {}
INFO  ... [shard 0:stmt] replication_strategy - prepare_options() {"replication_factor": "3"}
INFO  ... [shard 0:stmt] replication_strategy - network_topology_strategy() with {"datacenter1": "3"}
INFO  ... [shard 0:stmt] replication_strategy - validate_options() {"datacenter1": "3"}
INFO  ... [shard 0:stmt] migration_manager - Create new Keyspace: KSMetaData{name=ks, strategyClass=org.apache.cassandra.locator.NetworkTopologyStrategy, strategyOptions={"datacenter1": "3"}, cfMetaData={}, durable_writes=true, tablets={"initial":0}, userTypes=org.apache.cassandra.config.UTMetaData@0x600000082ba8}
INFO  ... [shard 0:stmt] replication_strategy - network_topology_strategy() with {"datacenter1": "3"}
INFO  ... [shard 0: gms] schema_tables - Creating keyspace ks
INFO  ... [shard 0: gms] replication_strategy - network_topology_strategy() with {"datacenter1": "3"}
INFO  ... [shard 1: gms] replication_strategy - network_topology_strategy() with {"datacenter1": "3"}
```

First time n.t.s. is constructed with empty options, I assume it's some temporary object that's needed to get I don't know what, maybe fully-qualified r.s. name. What's essential here is:
1. `prepare_options()` is called. If options have `replication_factor`, it will be fixed it into `dc` one. It may as well check for this option case and throw if it's not good
2. `foo_topology_strategy` constructor is called with "fixed" options
3. `foo_topology_strategy::validate_options()` is called with options fixed _and_ with initial_tablets set by ks_prop_defs::prepare_options() code, so it can check if tablets were requested and throw if not supported

Reply:
> Still this doesn't answer my question. The network_topology_strategy constructor has explicit check

Hmm I assumed your Q is: "How come the ks_prop_defs.cc::prepare_options() is called later?", to which I gave the stacktrace indicating it's called later, sorry, then I am lost. 

> First time n.t.s. is constructed with empty options, I assume it's some temporary object that's needed to get I don't know what, maybe fully-qualified r.s. name.

Yeah, probably, I see it too.
>  What's essential here is:
> 1. `prepare_options()` is called. If options have `replication_factor`, it will be fixed it into `dc` one. It may as well check for this option case and throw if it's not good

It could, but somehow checking for the tag's presence and unrolling it has been implemented in 2 separate places and because I'm not sure why, I left it like that.

> 2. `foo_topology_strategy` constructor is called with "fixed" options
> 3. `foo_topology_strategy::validate_options()` is called with options fixed 

Yes

> _and_ with initial_tablets set by ks_prop_defs::prepare_options() code, so it can check if tablets were requested and throw if not supported

No, because tablets are set only for n.t.s.

Reply:
> No, because tablets are set only for n.t.s.

Let's remove this check from ks-prop-defs and let it initialize initial tablets for any r.s., so that it could check it later in its validate_options()

Reply:
> > No, because tablets are set only for n.t.s.
> 
> Let's remove this check from ks-prop-defs and let it initialize initial tablets for any r.s., so that it could check it later in its validate_options()

I already had it implemented this way [here](https://github.com/scylladb/scylladb/commit/5c41d421f1e3679c82c0cdf0504645b603097126#diff-84071447808e8d409c99f3a0f556a5ff7894f3b182df67bbe03bd94d7165f2c1R44)*. And because of the [backward compatibilty problem](https://github.com/scylladb/scylladb/pull/20088#discussion_r1725071311) I had to reimplement it. I don't recall the exact reason why I couldn't stick with this approach. I vaguely remember it had something to do with being "too late" in the statement processing path, i.e. the place where I was setting tablets options (the 1st link should point directly to this line of code) lacked the context of how the tablets were set: only as a config option or via CQL, which is a crucial differentiation between allowing e.g. simplestrategy in CREATE KS with tablets turned on or rejecting it.

\* It's just a single commit out of the whole patches. I am not sure how to get the whole patchset, as I force-pushed changes + rebasing onto the latest `master`, and the diff contains both my changes and rebase, i.e. it's unreadable. But this commit gets the essence of it. BTW. lesson learned, always do the rebase separately.

Reply:
I took a look again and I don't think it's going to work, at least not the way that won't cause a bigger refactor. The problematic place in my previous implementation is this part:
```diff
- , _my_type(my_type) {}
+ , _my_type(my_type)
+ , _uses_tablets(params.initial_tablets.has_value()) {}
```
and specifically it is `params.initial_tablets`, which is of `std::optional<unsigned>` type and lacks the context of how tablets have been enabled - through configuration or CQL - which is crucial if we want to allow to CREATE KEYSPACE with e.g. simplestrategy when `enable_tablets` is set in the config but not in CQL, and in such case this option would be set to false, otherwise to true and we'd reject a ks that has tablets enabled in CQL but doesn't support them in the validation phase. However, this optional is engaged in case either config has tablets enabled or they've been enabled via CQL, which makes this value useless in the validation part, because of lack of this differentiation. The "bigger refactor" I mentioned earlier would mean to refactor this optional to resemble the struct I introduced, which provides the context of how tablets are enabled:
```
    struct init_tablets_options {
        bool enabled;
        bool enabled_via_cql;
        std::optional<unsigned> specified_count;
    };
```
but since this optional is already used in so many places, this would cause this bigger refactor. 

> so you setup _uses_tablets "unconditionally", i.e. -- regardless of whether specific r.s. supports it or not

Well, I was setting it in `abstract_replication_strategy::abstract_replication_strategy`, which is a base class of all replication strategies, and all of them have this member already. It was natural to set it if tablets were set either via config or cql and then return an error in the validation phase if a strategy has it set but doesn't support tablets. It is the current design that is broken, because all strategies have this `_uses_tablets` member, but only n.t.s. uses it really, so all strategies already know about tablets. And they should if in some cases we allow to create ks with all kind of strategies with tablets enabled.

> so only r.s. that inherits from tablet_aware_replication_strategy can turn it to true

So if somebody creates a ks with simplestrategy and enables tablets via CQL, the code representation of the replication strategy wouldn't have `_uses_tablets` set and we'd allow to create this ks, whereas we should return an error.